### PR TITLE
WIP: feat: Integrated passport.js for authentication

### LINKED ISSUE #1120: Changing towards passport.js for authentication
Lots of different OAuth providers are being requested, e.g. in #755, #962, or some other issues.

I'm in the process of working for #962 and was wondering about moving towards passport.js, which would make for a much easier provider addition procedure.

### PR DESCRIPTION
**Pull request by jakoblorz**
Following #1120, here is a WIP to integrate [`passport.js`](http://www.passportjs.org/). 

passport.js supports "sessionless" authentication. By doing that, passport.js acts as a boarding library handling the data-exchange, so the currently login, logout and session processes can be kept as they are. No additional cookies are introduced as well.
If we were to use passport.js like it is shown in most examples, it would act more as a complete serverside state manager, which would require additional changes not only on the serverside but also on the client. E.g. the `accessToken` cookie would be replaced by a `koa-session` managed cookie.

**If you want to support this PR, please have a look at https://github.com/outline/outline/pull/1183#issuecomment-626354357**

### Changes

- [x] Integrate `passport.js` into the Koa stack. Preferably without adding additional cookies or changing current login-flows
- [x] Migrate the Slack OAuth2 flow to the passport module. @MatheusRV suggested [passport-slack](http://www.passportjs.org/packages/passport-slack/)
- [x] Migrate the Google OAuth2 flow to the passport module

Depending on the scope set by @tommoor for this PR:
- [x] Dynamic strategy loading from `./services` or other folder/package.
- ~Database adjustment to replace vendor specific fields (such as `slackId` etc) with generalized fields. A field such as `service` already identifies the vendor.~

### Changelog
- `server/index.js`: Add [`koa-passport`](https://www.npmjs.com/package/koa-passport) to the stack.
- `server/utils/authorize.js`: Create wrapped `passport.authorize` to capture authentication results in an "awaitable" way
- `server/auth/slack.js`: Configure `passport-slack` to be used as an authentication strategy. Migrate `/slack` fully and `/slack.callback` partially to be handled by passport.

### Possible Features / Changes in this or following PRs
- Clean OAuth2 Interfaces / APIs, which would enable users to define (maybe even dynamically-loaded) modules for authentication (like in the `services/` package). https://github.com/outline/outline/pull/939#issuecomment-487390892
- Generally, passport.js enables third-party providers to authenticate users, mostly without custom code integration required. At this point, the third-party provider provides information about teams so any OAuth2 provider with some kind of team association would work: [Discord](https://discordapp.com/developers/docs/topics/oauth2) #729, [Mattermost](https://docs.mattermost.com/developer/oauth-2-0-applications.html) #1177, [Mastodon](https://docs.joinmastodon.org/methods/apps/oauth/), [GitHub](https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/) / [GitLab](https://docs.gitlab.com/ee/api/oauth2.html) #962, [Basecamp](https://github.com/basecamp/api/blob/master/sections/authentication.md) and OpenID, LDAP / Active Directory / Microsoft Teams #755 #1067 for corporations.

### Problems
- [x] [passport-slack](http://www.passportjs.org/packages/passport-slack/) seems to omit the state parameter (the query ends with `...&state=`), which is not optimal. I looked into the implementation of `passport-slack`, it is based on `passport-oauth` with only a very slim layer around. Interestingly, `passport-oauth` correctly handles state parameters, so it seems that this is lost in `passport-slack`. I believe that this should be resolved before moving ahead, either by writing a `passport-slack` implementation using `passport-oauth` or switching the package. What do you think @tommoor? I have no problem with reimplementing `passport-slack` in this PR. Edit: This problem will be resolved once I deliver/commit a `StateStore`, which `passport-oauth` requires. 

closes #1120

### CODE REVIEW

**Comment by jakoblorz** [comment]
Just to validate my passport.js integration, I implemented a dirty Discord integration on my fork: 7e4e7a9ba3d6e59819d2c259b5239be2bcce700d

The idea for that came from #729 (@muit @jasonquiz). Because one Discord user can be part of multiple guilds (teams), I went for the first one, which was created by the user logging in. This is just a test to see if the current passport.js integration does bring the expandability it advertises.

**Comment by tommoor** [comment]
Hey @jakoblorz – thanks for taking this on, I'm totally into making the authentication more plug-n-play – a lot of the auth code is some of the first that was written in the project. 

> passport.js supports "sessionless" authentication. By doing that, passport.js acts as a boarding library handling the data-exchange

Great 👍 

>  Database adjustment to replace vendor specific fields (such as slackId etc) with generalized fields. A field such as service already identifies the vendor.
>  Dynamic strategy loading from ./services or other folder/package.

I'd leave these for separate follow up PR's, that's going to be the best way to make this landable on master in a reasonable timeframe 😄 

> What do you think @tommoor? I have no problem with reimplementing passport-slack in this PR. Edit: This problem will be resolved once I deliver/commit a StateStore, which passport-oauth requires.

I'm actually unsure what you mean by the final sentence here, but yes we definitely don't want to regress the security of including a state param. Considering how old that package is and slim it makes sense to use it as a base for an in-house strategy…

**Comment by jakoblorz** [comment]
>>    What do you think @tommoor? I have no problem with reimplementing passport-slack in this PR. Edit: This problem will be resolved once I deliver/commit a StateStore, which passport-oauth requires.

> I'm actually unsure what you mean by the final sentence here, but yes we definitely don't want to regress the security of including a state param. Considering how old that package is and slim it makes sense to use it as a base for an in-house strategy…

Okay, I'll rephrase: I took another look at the passport-oauth package, it seems that it needs a _StateStore_ which is responsible for storing the state and later verifying it. The ones that come with the package are using the express-session middleware, which is unavailable for use via the koa-router. I believe that using koa-session could work, but again, that would introduce another layer of complexity for session management. Using the [reference _StateStore_](https://github.com/jaredhanson/passport-oauth2/tree/master/lib/state), I would go for implementing one myself, reusing the code that is currently in master and is [responsible for setting](https://github.com/outline/outline/blob/master/server/auth/slack.js#L27) the state cookie. `store()` would then set the state cookie and `verify()` would verify and the remove it.
Because passport-slack passes the options parameter through to passport-oauth, I think it would be possible to use the custom _StateStore_ with passport-slack. But again, because you already support me in building a in-house strategy for slack, this pass-through idea is not that important.


>> Database adjustment to replace vendor specific fields (such as `slackId` etc) with generalized fields. A field such as service already identifies the vendor.
    Dynamic strategy loading from `./services` or other folder/package.

> I'd leave these for separate follow up PR's, that's going to be the best way to make this landable on master in a reasonable timeframe :smile:

Okay, to sum up, the scope for this PR is ...
... Introduce passport.js
... Convert existing auth routes to passport.js, including in-house strategy implementation for slack, maybe also google

Then, I will add 2 additional PRs:
1. Update database models to generalize current vendor-specific fields
2. Auto-mount system for dynamic strategy loading, similar to `./services`, including proper interfacing / abstraction so that integration of custom strategies can be achieved in a short timeframe

**Comment by tommoor** [comment]
> I think it would be possible to use the custom StateStore with passport-slack

If we're going to need it anyway for future auth methods perhaps it's better to just go ahead and implement that rather than bring passport-slack internal?

**Comment by tommoor** [comment]
> Update database models to generalize current vendor-specific fields

Heads up that isn't going to be as easy as it sounds as we currently support a team having both `slackId` and `googleId`.

**Comment by jakoblorz** [comment]
> > I think it would be possible to use the custom StateStore with passport-slack
> 
> If we're going to need it anyway for future auth methods perhaps it's better to just go ahead and implement that rather than bring passport-slack internal? Sounds like you'll need it either way

I commited the StateStore with 4e207bd already, so no worries on that front. Because it is not necessarily tied to this project, I am thinking about packaging it for reuse in other projects and introduce it as another dependency. Is that okay with you? I am sure there are more people out there who have combined koa and passport and may not even realize that the state is not being set.

> > Update database models to generalize current vendor-specific fields
> 
> Heads up that isn't going to be as easy as it sounds as we currently support a team having both `slackId` and `googleId`.

Okay, wasn't aware of that. That makes the problem interesting, I'm sure I will come up with something.

**Comment by tommoor** [comment]
>  I am thinking about packaging it for reuse in other projects and introduce it as another dependency. Is that okay with you?

No problem at all, as long as the version is pinned.

**Comment by zdianjiang** [comment]
@jakoblorz  

I have clone your branch and `make up`, But have some errors. these is my process..
Also I have do `git rebase upstream/master` 

1. bash error
```bash
docker-compose up -d redis postgres s3
jakoblorz_outline_redis_1 is up-to-date
jakoblorz_outline_s3_1 is up-to-date
jakoblorz_outline_postgres_1 is up-to-date
docker-compose run --rm outline bash -c "yarn && yarn sequelize db:migrate"
Starting jakoblorz_outline_postgres_1 ... done
Starting jakoblorz_outline_s3_1       ... done
Starting jakoblorz_outline_redis_1    ... done
internal/modules/cjs/loader.js:985
  throw err;
  ^

Error: Cannot find module '/opt/outline/bash'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.Module._load (internal/modules/cjs/loader.js:864:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)
    at internal/main/run_main_module.js:18:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
make: *** [up] Error 1
```
I have to edit Makefile and change `bash` to `/bin/sh` 
```Makefile
up:
	docker-compose up -d redis postgres s3
	docker-compose run --rm outline /bin/sh -c "yarn && yarn sequelize db:migrate"
	docker-compose up outline
```

2. error outline@0.2.0: The engine "node" is incompatible with this module
```bash
docker-compose up -d redis postgres s3
jakoblorz_outline_redis_1 is up-to-date
jakoblorz_outline_postgres_1 is up-to-date
jakoblorz_outline_s3_1 is up-to-date
docker-compose run --rm outline /bin/sh -c "yarn && yarn sequelize db:migrate"
Starting jakoblorz_outline_s3_1 ... done
Starting jakoblorz_outline_postgres_1 ... done
Starting jakoblorz_outline_redis_1    ... done
yarn install v1.22.0
[1/5] Validating package.json...
error outline@0.2.0: The engine "node" is incompatible with this module. Expected version "8.11". Got "12.16.1"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
make: *** [up] Error 1
```
I have to edit Makefile and add `yarn config set`
```Makefile
up:
	docker-compose up -d redis postgres s3
	docker-compose run --rm outline /bin/sh -c " yarn config set ignore-engines true && yarn && yarn sequelize db:migrate"
	docker-compose up outline
```

3. problem also

```bash
yarn run v1.22.0
$ /opt/outline/node_modules/.bin/sequelize db:migrate

Sequelize [Node: 12.16.1, CLI: 2.8.0, ORM: 4.28.6]

gulp[101]: ../src/node_contextify.cc:649:static void node::contextify::ContextifyScript::New(const v8::FunctionCallbackInfo<v8::Value>&): Assertion `args[1]->IsString()' failed.
Done in 8.60s.
docker-compose up outline
jakoblorz_outline_redis_1 is up-to-date
jakoblorz_outline_postgres_1 is up-to-date
jakoblorz_outline_s3_1 is up-to-date
Creating jakoblorz_outline_outline_1 ... done
Attaching to jakoblorz_outline_outline_1
outline_1   | yarn run v1.22.0
outline_1   | error outline@0.2.0: The engine "node" is incompatible with this module. Expected version "8.11". Got "12.16.1"
outline_1   | error Commands cannot run with an incompatible environment.
outline_1   | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
jakoblorz_outline_outline_1 exited with code 1
```

**Comment by tommoor** [comment]
It seems like somehow you've tried to merge the branch with an incredibly old version of master, perhaps years old.

**Comment by zdianjiang** [comment]
> It seems like somehow you've tried to merge the branch with an incredibly old version of master, perhaps years old.

Yes I have see also, I think the passport-js-integration branch is base on the old version...
Cant see it in `git log`

**Comment by gempain** [comment]
@jakoblorz any timeline on this ?

**Comment by jakoblorz** [comment]
@gempain had a full plate these last weeks but still intent to land the PR. Will push the changes in the next 2 weeks.

**Comment by gempain** [comment]
@jakoblorz looking forward it !

**Comment by almereyda** [comment]
Is there anything one could do to support here, to keep this going a little forward?

**Comment by gempain** [comment]
Yep we're definitely looking forward to this for dropping WikiJS. We'd be happy to provide some help if needed !

**Comment by gempain** [comment]
We'd be happy to provide some help if needed !

**Comment by jakoblorz** [comment]
So I will proceed with testing my implementation with the next few commits. I do need a someone / few people (the more the merrier) who is / who are willing to test this. Most importantly, @gempain @almereyda if you could test and / or implement the google authentication, I don't have a google business login.

For help or usage:
- I have written a [guide](https://github.com/jakoblorz/outline/blob/passport-js-integration/server/auth/README.md#building-custom-authentication-providers) on how to implement other providers, so please take a look if you are up to it. There are some spelling errors in there for sure, I will also accept PRs for them.
- There is also a reference implementation for discord in https://github.com/jakoblorz/outline/blob/passport-js-integration/server/auth/discord.js#L67
- I have also ported the slack routes already: https://github.com/jakoblorz/outline/blob/passport-js-integration/server/auth/slack.js#L39

Generally, I will be accepting PRs on my fork, so that we have a clean upstream commit log in this PR. @zdianjiang was right, I was on a older version initially, but that seems to be corrected by now.

**Comment by gempain** [comment]
@jakoblorz thanks for the awesome effort ! I will work on the Mattermost integration and I'll get back to you after testing it out.

**Comment by almereyda** [comment]
Thank you for pushing this forward. I am able to check against Google, GitHub and GitLab within the week.

**Comment by gempain** [comment]
@jakoblorz I have it a shot tonight. I can get the Mattermost integration working, but I got an error the first time I tried, it was a page showing a JSON body which contained the OAuth code. I am not able to reproduce since, so I'll try tomorrow with a clean DB. Two comments so far:
1. Mattermost's `avatarUrl` is only accessible if you send an auth header with the request, so I had to set it to `''` as done with DIscord. After login, the organization/user avatar doesn't load, as expected, but instead shows the image `alt`. It's quite ugly and I think we should either hide it or use a gravatar url. Details, but details matter.
2. I'm not familiar with the error handling in Outline, so I was wondering what we're supposed to use as second parameter of the `customError` function ?
    ```js
    class TooManyTeams extends customError(
      'TooManyTeams',
      'notice=auth-error&error=state_mismatch' // TODO not sure what to put here
    ) {}
    ```
3. It's not clear how to handle when a user has several teams. In Discord, you pick the first or first owned by the user when he's got more than one. Otherwise, you throw an error. For Mattermost, I just throw an error when you don't have exactly one team. That should fit most cases.

**Comment by jakoblorz** [comment]
> @jakoblorz I have it a shot tonight. I can get the Mattermost integration working, but I got an error the first time I tried, it was a page showing a JSON body which contained the OAuth code. I am not able to reproduce since, so I'll try tomorrow with a clean DB. Two comments so far:
> 
> 1. Mattermost's `avatarUrl` is only accessible if you send an auth header with the request, so I had to set it to `''` as done with DIscord. After login, the organization/user avatar doesn't load, as expected, but instead shows the image `alt`. It's quite ugly and I think we should either hide it or use a gravatar url. Details, but details matter.

Have a look at the pending Google OAuth2 integration. It creates 2 different avatars using tiley or clearbit.
https://github.com/jakoblorz/outline/blob/passport-js-integration/server/auth/google.js#L64

> 2. I'm not familiar with the error handling in Outline, so I was wondering what we're supposed to use as second parameter of the `customError` function ?
>    ```js
>    class TooManyTeams extends customError(
>      'TooManyTeams',
>      'notice=auth-error&error=state_mismatch' // TODO not sure what to put here
>    ) {}
>    ```

These `notice=auth-error&error=state_mismatch` are the redirect query parameters in case of failure. You can use them in the error handler so that one kind of error is defined along its redirect url. Have a look at [google.js](https://github.com/jakoblorz/outline/blob/passport-js-integration/server/auth/google.js) once again. The contents of this redirect are hardcoded in the server-side rendered app and will resurface in [AuthNotices.js](https://github.com/outline/outline/blob/master/server/pages/components/AuthNotices.js)

> 3. It's not clear how to handle when a user has several teams. In Discord, you pick the first or first owned by the user when he's got more than one. Otherwise, you throw an error. For Mattermost, I just throw an error when you don't have exactly one team. That should fit most cases.

Yes, you are right. The discord integration is just a kind of playground, I used it to test different possible ways to integrate. This is reminiscent of that time. Further down the line I am thinking of enabling integration developers with the tools to show an intermediate page to the user. In this use case, the users would then be able to select the correct team. On that note, I will also expose methods so that the developer can customize the login-with XY button contents. Btw, you wrote in #1269 _@jakoblorz is done with the PR_. As you can see here, there are more changes incoming, because we want to integrate as many different use-cases. Handling multiple teams is just one edge-case. To ensure correct integrations, there will be extensive testing, additions of UI-oriented options and possible dynamic, branch oriented service integrations added.

**Comment by gempain** [comment]
@jakoblorz what I really meant was that most of the work was quite forward already. I'll revise my definition of done 😛

**Comment by MatheusRV** [comment]
LGFM

**Comment by gempain** [comment]
@jakoblorz @tommoor just wanted to kick in to see where we're at here ? Do we have an estimate timeline for this PR to be merged in ?

File: server/auth/google.js [hunk 1]
```diff
@@ -1,57 +1,55 @@
...
+  accessToken,
+  refreshToken: string
+): Promise<DeserializedData> {
+  const profile = await json('https://www.googleapis.com/oauth2/v1/userinfo', {
+    headers: {
+      authorization: `Bearer ${accessToken}`,
+    },
   });
 
   if (!profile.data.hd) {
```
Comment (almereyda) [inline server/auth/google.js:45]: GoogleHDError is never thrown when trying to login with a regular Google account, since the check for the condition cannot be performed, as it just happened to me.

```
TypeError: Cannot read property 'hd' of undefined
    at _callee3$ (/opt/outline/server/auth/google.js:45:21)
    at tryCatch (/opt/outline/node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:65:40)
    at Generator.invoke [as _invoke] (/opt/outline/node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:303:22)
    at Generator.prototype.<computed> [as next] (/opt/outline/node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:117:21)
    at step (/opt/outline/server/auth/google.js:177:191)
    at /opt/outline/server/auth/google.js:177:361
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by gempain** [comment]
> Thanks to the nice example of @gempain in [gempain@7ffd600](https://github.com/gempain/outline/commit/7ffd600c9962c064da78b6f89a2390724a97f9bb) I was able to manufacture a working GitLab strategy based off this PR. I suspect a GitHub one would look very similar.

Thanks a lot for the kind words !

**Comment by jakoblorz** [comment]
@gempain 

> @jakoblorz @tommoor just wanted to kick in to see where we're at here; do we have an estimate timeline for this PR to be merged in ?

I for myself am very packed at the moment. I do want to see this PR landing as soon as possible though. I think I will be able to continue working on it after my final bachelors exams at the end of August.

I do see that #1317 is kind of blocking the current implementation. I don't think that it makes sense to continue implementing on the current state, fully aware that the authentication part, which is the central part, will change in the future. I personally would like to wait for #1317 to land. #1317 kind of solves the issue that I raised in the very first comment above:

> ~Database adjustment to replace vendor specific fields (such as slackId etc) with generalized fields. A field such as service already identifies the vendor.~

After #1317 landed, I would then proceed with
- [ ] fixing up the strategy implementation, possibly with a more intuitive interface / documentation
- [ ] fixing issues in the provided strategies @almereyda
- [ ] provide a way to integrate frontend pages into the strategies (for provider specific on-boarding processes) as well as a custom "login with" button, which the landing page will render depending on the registered strategies.

**Comment by gempain** [comment]
@jakoblorz awesome ! Wishing you good luck with your bachelors ! Let's wait for #1317 to be merged then, it does make sense. Looking forward to all this :) Thanks for the feedback and for your help !

**Comment by almereyda** [comment]
Thank you for the details about upcoming progress, and best of luck and concentration for those exams, indeed!

This gives us the opportunity to be patiently preparing our communities for the landing of this beautiful extension of the Outline ecosystem.

For then we can approach deeper integration with Mattermost, Element/Matrix, similar to what exists for Slack already. Eventually that would mean also abstracting and generalising the bot interface, in so different implementations could make use of generally available features.

**Comment by MatheusRV** [comment]
Friendly ping

**Comment by jakoblorz** [comment]
> Friendly ping

Ping received, I will be on it 😁

**Comment by rhummelmose** [comment]
This would be awesome, currently blocking me from using Outline in a team.

**Comment by awinogradov** [comment]
Hi, guys! So awesome PR. Can't wait to see it in master ;) Any plans to merge and release this beauty?

**Comment by razor950** [comment]
is this still happening? I'd love to try outline but not have to use it with slack and it seems this would provide that if I am understanding correctly.

**Comment by rhummelmose** [comment]
Nothing on this since July. Can we expect this to be merged at any point in time?

**Comment by zdianjiang** [comment]
hello @tommoor , when time we can merge this PR . What can i do for this asap ?

**Comment by MatheusRV** [comment]
#most-required-feature

**Comment by jakoblorz** [comment]
As I see it, #1317 is kind of blocking this effort (see https://github.com/outline/outline/issues/1317#issuecomment-707215506). 
Sure we could polish and push this, but once #1317 is implemented, we would have to change this implementation again. And by doing that, we would also break any effort made by you, the community, to implement custom providers. The issue I am seeing here was also mentioned in a broader sense by @almereyda in https://github.com/outline/outline/pull/1183#pullrequestreview-449125696. 
> It is left here as a basis for discussion about how we would like to proceed with preparing further strategies, in terms of possibly multiple group-team-mappings, or if we deem a strategy to be bound to a single group only, as currently done in the Mattermost and my example.
Having an explicit team authentication table is a start to allow multiple group-team-mappings etc.

So: Does someone feel comfortable enough to implement #1317? @gempain @Bloodb0ne @almereyda @MatheusRV I really don't have the time to wrestle with database migrations and possibly changing huge chunks of controller code. (@tommoor I know you are doing many things at many places. But as the operator of the SaaS offering, it might be wise to at least supervise the database migrations?) Am I overestimating the complexity of the issue here? I am still up for this pull though.

**Comment by tommoor** [comment]
> @tommoor I know you are doing many things at many places. But as the operator of the SaaS offering, it might be wise to at least supervise the database migrations?

Yep, I will do that if any formal PR's are put in – I also outlined the kind of thing I'd expect to see in that issue. In terms of timeline from myself this probably comes after multiplayer collaboration as the next "big ticket" item.

**Comment by jdaviescoates** [comment]
Any chance of getting this merged? Or are we waiting for @jakoblorz to resolve the conflicts?

It'd be great to be able to finally use Outline without corporate logins!

**Comment by Vetm** [comment]
Any news on this?

**Comment by razor950** [comment]
what is needed to progress on this?

**Comment by bglidwell** [comment]
Also very excited to see this coming.

**Comment by paambaati** [comment]
@tommoor If you don’t mind me asking, is this closed by mistake or is this not in your roadmap? I looked at recent commits & PRs and I don’t see this feature landed.

**Comment by tommoor** [comment]
I didn't actually close the issue, I deleted the `master` branch as it was confusing after our move to `main` several weeks ago. Apparently it automatically closed this PR instead of rebasing it, I'm not sure why to be honest.

**Comment by jeffmcneill** [comment]
> I didn't actually close the issue, I deleted the `master` branch as it was confusing after our move to `main` several weeks ago. Apparently it automatically closed this PR instead of rebasing it, I'm not sure why to be honest – many other PR's successfully transferred.
> 
> Improving the available authentication methods is still on the roadmap and still blocked behind this big migration: #1317, sorry for any confusion caused.

So, should this issue be re-opened or is it duplicated somewhere else? Without the authentication change, I'm unable to use this.

**Comment by almereyda** [comment]
This PR has been superseded by #1934, for those who missed the notification.

**Comment by matbrgz** [comment]
For those are interested in start PR in this repository a good place to start is https://github.com/outline/outline/issues/817
